### PR TITLE
feat: support mapping runtype name to something else in the output

### PR DESCRIPTION
**Pull request by rkok**
Adds support for renaming the runtypes to something like this, similar to how [json-to-runtypes](https://github.com/runeh/json-to-runtypes#readme) does it:

```ts
export const FooTypeRt = Null.Or(String);
export type FooType = Static<typeof FooTypeRt>;

export const HorseTypeRt = Record({ a: FooTypeRt, b: FooTypeRt, });
export type HorseType = Static<typeof HorseTypeRt>;
```

by doing:

```ts
const generator = new Generator({
  // ...
  mapRuntypeName: (type) => `${type}Rt`, // Suffix all resulting runtypes with "Rt"
})
```

Not sure if I really like the use of the term 'map' here, but that's the best I could come up with. Curious to hear your thoughts

### CODE REVIEW

**Comment by johngeorgewright** [comment]
Thanks @rkok.

I was wondering if this could be used in configuration too. Something like:

```yml
- targetFile: foo.runtypes.ts
  sourceTypes:
    # ...
    mapRuntypeName: ${name}Rt
    mapTypeName: ${name}T
```
 
This obviously would require a templating language, but happy with that as it'll keep consistency.

And, if we used this approach, it would make sense to use the same format programatically:

```typescript
const generator = new Generator({
  // ...
  mapRuntypeName: '${type}Rt',
})
```

What do you think?

**Comment by johngeorgewright** [comment]
Or, maybe it's more sensible not to allow overriding of names, but rather configuration options for "prefixes" and "suffixes"?

```yaml
- targetFile: foo.runtypes.ts
  sourceTypes:
    - runtypeSuffix: Rt
      typePrefix: Type_
```

**Comment by rkok** [comment]
I quite like your original idea of a single 'format string' argument for the runtype and the TS type. We could do something like:

```yml
- targetFile: foo.runtypes.ts
  sourceTypes:
    runtypeFormat: Prefix{type}Suffix
    typeFormat: Prefix{type}Suffix
```

(and equivalent for programmatic use)

And then do:

```ts
// somewhere.ts
const outputRuntype = config.runtypeFormat.replace('{type}', inputRuntype);
const outputType = config.typeFormat.replace('{type}', inputType);
```

Let me update the PR sometime later this week :-)

**Comment by rkok** [comment]
@johngeorgewright The PR is updated. Please let me know what you think, also regarding the structuring of it all - I'm eager to learn.

**Review by johngeorgewright — APPROVED** [review]
This looks good for now. I may, in the future, remove the requirement for 2 specific `TypeNameFormatter` constructor parameters and make it more generic. For example,

```typescript
this.#typeFormatter = typeFormatter(typeFormat)
this.#runtypeFormatter = typeFormatter(runtypeFormat)
# ...
const typeName = this.#typeFormatter(originalTypeName)
const runtypeName = this.#runtypeFormatter(originalRuntypeName)
```

This way the formatter could be simplified a lot: (crude example)

```typescript
export default function typeFormatter(typeFormat?: string) {
  return (name: string) =>
    typeFormat ? typeFormat.replace(TYPE_PLACEHOLDER, originalName) : name
}
```

But that's probably just more of a personal preference and it isn't necessary for now.

Thanks! I'll release this soon.

File: src/Generator.ts [hunk 1]
```diff
@@ -147,13 +157,15 @@ export default class Generator {
     sourceType: string,
     sourceImports: Set<string>
   ) {
+    let sourceTypeRt = this.#nameFormatter.formatRuntypeName(sourceType)
```
[the remark this answers is not in the document]
Comment (rkok) [reply src/Generator.ts:160]: Cheers for the feedback, valid points.
Anyway, happy it's merged :-)

**Pull request merged**